Proposal: Unit type checking

michaeldickens opened the pull request:
Add unit type checking to Squiggle.

## Syntax

Squiggle unit types are declared using approximately the same syntax as TypeScript. Example:

    x = 1  // implicit type
    y :: kg = 1  // type is `kg`

A type signature consists of one or more identifiers separated by zero or more `*` symbols, followed by zero or more identifiers separated by zero or more `/` symbols. In other words, type signatures must match the pseudo-regex `\identifier(*\identifier)*(/\identifier)*`.

Some more examples of type signatures:

    speed :: m/s = 1
    force :: kg*m/s/s = 2
    work :: kg*m*m/s/s = 3

## Semantics

The parser type-checks variable definitions and expressions.

For this demo, the type checking rules are fairly simple:

- Any variable without a declared type is simply assumed to be the correct type, even if that's not logically possible.
- An expression that multiplies or divides two variables with known types infers the type of the result. For example, if `x` has type `kg` and `y` has type `liter`, `x/y` has type `kg/liter`.
- If two values with known types are combined using an infix operator other than `*` or `/`, the type checker enforces that the two values have the same type. For example, you can only add/subtract two variables if they have the same type (or if at least one of them does not have a declared type).
- If a typed variable is assigned to a value with a known type, the type checker enforces that the value has the correct type, and raises a type error if not.

## Implementation details

I implemented type checking in the way that looked best to me, but Squiggle maintainers may have a better idea of how to do it.

In [parse.ts:parse()](https://github.com/michaeldickens/squiggle/blob/type-checking/packages/squiggle-lang/src/ast/parse.ts#L35), after generating the AST, I added a call to the `typeCheck()` function, which is defined in a new file [typeChecker.ts](https://github.com/michaeldickens/squiggle/blob/type-checking/packages/squiggle-lang/src/ast/typeChecker.ts).

## Open questions

There's definitely more that needs to be implemented to make this PR viable, but I didn't want to make too many design decisions without consulting the Squiggle team.

1. Should the type checker do type inference, or should it simply ignore any variable without a type (like TypeScript's `any`)? Type inference is more work but it looks doable to me. But it's not obvious that type inference is actually desirable from a design standpoint. (I lean toward yes.)
2. How to treat numeric literals? I see two reasonable possibilities: (1) let them be any type, or (2) treat them as unitless unless directly assigned to a variable (e.g. in `x: kg = 1`, the `1` should be treated as having type `kg`, not unitless). Treating them as unitless would help catch more type errors, but it also means you'd need a syntax for type conversion. For example, if you have a distance literal and a time variable, converting them to velocity requires declaring the type of the literal, such as `usainBoltSpeed: m/s = (100 as m) / raceTime`.
3. Is my proposed syntax ok? There are ways it might be improved, for example allowing parentheses and/or exponents:
       work: (kg*m^2)/(s^2) = 1
4. Do probability distributions require any special treatment? I believe they can be treated just like numeric values.
5. What syntax should function definitions use? I'm thinking it could use basically the same syntax as TypeScript:
       func(x: type): type = ...
6. How to type arrays and dicts? Should Squiggle require that the values in an array/dict all have the same type? What should the type signature syntax be? Again, it can probably be the same as TypeScript.
7. It would be good for SquiggleHub to have a good way of showing types. How should `squiggle-lang`  supply type info make that work? (SquiggleHub does some stuff with GraphQL that I don't understand but I haven't looked into it much.)
8. How to handle type errors? Should they be a subclass of `ICompileError`, or should they be handled separately? Does the frontend need some special way to consume them?

michaeldickens commented on packages/squiggle-lang/__tests__/ast/parse_test.ts:
```diff
@@ -219,8 +226,8 @@ describe("Peggy parse", () => {
       `
-  /* This is 
-  a multi line 
+  /* This is
```
my auto linter changed this, I didn't mean to include it in the PR

berekuk commented:
Quick comment based on the overall idea, I haven't looked at the implementation carefully yet.

Thanks for doing this! This is definitely something we discussed and want to support.

My main problem with the syntax that you picked is that besides unit type checking, we'll also want to implement other type checks in the future.

For this reason I'd strongly prefer if we used something else than `:` to provide the unit portion of the type.

This is easy to fix if we go with `::` instead of `:`, for now:

```
speed :: m/s = 1
```

The design for actual _types_ (that would allow to express other value types, e.g. functions or lists or whatever) is still not finalized and will probably take a while, but note that we already have [annotations in function signatures](https://www.squiggle-language.com/docs/Guides/Functions#parameter-annotations).

Units are obviously not types, because:
- they can't express any types other than numbers
- their names are too free-form, so you can't refer to type or variable names in type signatures

Btw, this is tangential, but my current plan is to eventually support domains everywhere:
- domains are values (can be constructed in runtime) that can be used in type signatures
- domains are generics from which compile-time types can be inferred
- domains also do runtime checks when compile-time checks are too difficult
- we could also have syntax sugar because full domain definitions are too clunky; similar to how in function signatures we already upgrade `[5, 10]` to `Number.rangeDomain(5, 10)`; units could also be extended in the similar way, `x :: m/s` -> `x: Number.unit("m") / Number.unit("s")`

With domains, we could start with completely runtime type-checks and gradually push the boundary towards compile-time checks where it's possible; but it won't be possible in all cases.

This also means that for now I don't really care if unit type checking or any other type checking is done in the compile-time. It's fine to prototype type checking in runtime, and then optimize later (if we pay attention and avoid features that are too powerful and can't be optimized later).

berekuk commented on packages/squiggle-lang/src/ast/parse.ts:
```diff
@@ -94,6 +96,22 @@ export function nodeToString(
       case "Decorator":
         return sExpr([node.name, ...node.args].map(toSExpr));
       case "LetStatement":
+        if (node.typeSignature.isImplicit) {
```
Too much copy-paste here, I recommend handling both types of statements together, and then inline-expand with:
```
[
  toSExpr(node.variable),
  node.type === "LetStatement" && !node.typeSignature.isImplicit
    ? toSExpr(node.typeSignature)
    : undefined,
  ...
]
```

(note that `sExpr` skips all undefined values)

michaeldickens replied:
I tried this but it didn't work because `nodeToString` puts spaces between each entry without skipping `undefined`, so what's supposed to be (say) "Identifier UnitValue" instead gets written as "Identifier  UnitValue" (with 2 spaces). I considered adding a separate pass to remove duplicate strings but that seemed messy.

berekuk replied:
Huh, are you sure? Maybe you tried `""` instead of `undefined`?

AFAICT `sExprToString` code filters out undefined values first.

michaeldickens replied:
Turns out this does work as you suggested, I'm not sure what problem I was having with it before

berekuk reviewed (changes requested):
The type checker, as you pointed out, is very incomplete.

Releasing incomplete prototypes is fine, but I'm somewhat concerned that future versions of Squiggle would handle more cases, which would make upgrading harder. Backward compatibility is very nice to have, and we try to auto-upgrade user models on Squiggle Hub whenever possible.

As an example,
```
x = {
  y : m = 1
  z : s = y
  z
}
```
would work with this PR but would break in the future.

Another example, as you mentioned, is type inference.

This might still be worth it, though, I'm honestly not sure.

Another option is just to do checks and inference in runtime; store unit information in [tags](https://www.squiggle-language.com/docs/Api/Tag). But this is a completely different approach from the one you use here; you'd have to:
- store type inference in compiled expressions
- make at least some basic builtin functions (the ones used for infix calls) aware of unit tags and infer them correctly

I think this might be easier and the functionality would be more complete; also it's clearer how to get from there to displaying unit information in the playground viewer, since it would be stored on values.

berekuk commented:
Now on these questions:

> 1. Should the type checker do type inference, or should it simply ignore any variable without a type (like TypeScript's any)? Type inference is more work but it looks doable to me. But it's not obvious that type inference is actually desirable from a design standpoint. (I lean toward yes.)

Type inference sounds very useful, and as I said above, its lack is dangerous because of backward compatibility problems.

I'm leaning towards "let's just infer in runtime", though.

> 2. How to treat numeric literals? I see two reasonable possibilities: (1) let them be any type, or (2) treat them as unitless unless directly assigned to a variable (e.g. in x: kg = 1, the 1 should be treated as having type kg, not unitless). Treating them as unitless would help catch more type errors, but it also means you'd need a syntax for type conversion. For example, if you have a distance literal and a time variable, converting them to velocity requires declaring the type of the literal, such as usainBoltSpeed: m/s = (100 as m) / raceTime.

> 3. Is my proposed syntax ok? There are ways it might be improved, for example allowing parentheses and/or exponents:
`work: (kg*m^2)/(s^2) = 1`

It's fine if we don't support these for now.

> 4. Do probability distributions require any special treatment? I believe they can be treated just like numeric values.

I'm not sure. I think units syntax is fine for dists if we store types separately (`x : type :: units = ...`, as I said in my previous comment), but I haven't thought about this too much.

> 5. What syntax should function definitions use? I'm thinking it could use basically the same syntax as TypeScript:
    func(x: type): type = ...
> 6. How to type arrays and dicts? Should Squiggle require that the values in an array/dict all have the same type? What should the type signature syntax be? Again, it can probably be the same as TypeScript.

I don't consider units to be full types, and doing both at the time would be too complicated; see my comment on domains above.

> 7. It would be good for SquiggleHub to have a good way of showing types. How should squiggle-lang supply type info make that work? (SquiggleHub does some stuff with GraphQL that I don't understand but I haven't looked into it much.)

GraphQL doesn't seem relevant here; runtime inference based on tags would be very helpful.

> 8. How to handle type errors? Should they be a subclass of ICompileError, or should they be handled separately? Does the frontend need some special way to consume them?

`ICompileError` seems fine for this, can be improved later. But, again, if we do runtime inference, then it's going to be just a normal runtime error, for now.

michaeldickens commented on packages/squiggle-lang/src/ast/peggyParser.peggy:
```diff
@@ -58,13 +58,38 @@ decorator
+  = ""
+    { return h.nodeImplicitType(location()); }
+
+typeWord
```
You're right, I was using it in an early draft and then forgot to delete it

michaeldickens commented on packages/squiggle-lang/src/ast/types.ts:
```diff
@@ -178,6 +180,7 @@ type LetOrDefun = {
 type NodeLetStatement = N<
   "LetStatement",
   LetOrDefun & {
+    typeSignature: NodeTypeSignature;
```
I think that would work, I figured it was bad practice to use null/undefined as an expected outcome which is why I created an explicit `ImplicitType` class. I could make it null instead. (Although the Squiggle style guide in the readme said to prefer `undefined` over `null`?)

berekuk replied:
Yeah, I recommended nulls here because there are nulls in `NodeFloat` types, and consistency is more important than the style guide.

But I don't really remember why I used `null` instead of `undefined` in `NodeFloat`, and it doesn't matter much.

michaeldickens commented:
@berekuk Based on your comments, here's my to-do list. Let me know if I should add anything.

1. Update unit type syntax to not conflict with (future) normal type syntax.
2. Implement type inference.
3. Treat numeric literals as any unit type. (Note: This will create backward compatibility issues if we decide later to make literals unitless.)
4. Add unit type syntax to the prettier plugin.
5. Unit type checking on distributions.
6. Minor fixes e.g. delete unused code.

Also: Can you explain what you mean by "domain"? And relatedly, why do you think it's too complicated to do unit types for function definitions and arrays/dicts?

berekuk commented:
> Update unit type syntax to not conflict with (future) normal type syntax.
> Implement type inference.

I think that tracking unit info in runtime, through tags, would be easier for now, and would give us same benefits as full type inference at the cost of a slight performance overhead. But either approach is fine, and we'll want compile-time type checks and inference eventually. I just think it's more work for you.

Full type inference is especially tough, so I'll explain my priorities for it:
1. some solution for units in nested blocks is a blocker for merging this PR (worst case scenario, just ban unit syntax in blocks entirely)
2. type inference on trivial assignments and arithmetic operations is also nice for the future backward compatibility, but I would agree to merge if (1) is fixed
3. type inference on generic functions is 

For clarity, in (1), I mean this:
```
x = {
  y :: kg = 1
  z :: m = y
  y + z
}
```

This would be quite unexpected for users, the feature looks like it should work but it doesn't do anything, and it's a footgun. So I'd like this to be supported before we merge.

In (2) I mean this:
```
x :: kg = 1
y = x
```

And this:
```
x :: m = 1
y :: s = 2
z = x / y    /* m/s? */
```

And in (3) I mean this:

```
f(a) = a
x :: kg = 1
y = f(x)
```

This one is the most difficult, and you definitely shouldn't attempt it now; some languages (Haskell and OCaml) are smart enough to infer that `f` maps any generic `'a` to `'a`, so it would infer that `y` has the same type as `x`. But this would require something like full [Hindley–Milner](https://en.wikipedia.org/wiki/Hindley%E2%80%93Milner_type_system).

Or, you could get all of (1), (2) and (3) for free, and also easier access to the unit information during render, if we do runtime tracking with tags.

berekuk commented:
> Also: Can you explain what you mean by "domain"?

The term "domain" is briefly mentioned in the [parameter annotation](https://www.squiggle-language.com/docs/Guides/Functions#parameter-annotations) documentation, but not very well explained.

The idea is that any check ("is this a valid value to use here?") can include runtime checks ("does the value belong to the [10, 20] range?") and compile-time checks ("is this a number?"). We plan to bundle both of these checks in a single "domain": a meta-value that describes both components.

Unlike in Typescript, domains are values, not a separate namespace of "type names". They can be constructed in runtime, e.g. with functions, passed around as parameters, and so on.

```
xDomain = Number.rangeDomain({ min: 5, max: 10 })
x: xDomain = 7
```

Here, `xDomain` type would be, let's say, `Domain<Number>`, and when you annotate `x: xDomain`, Squiggle would infer that `x` type would be `Number`. So it would do the compile-time check that `x` is a number, by checking that `7` is also a number.

In addition to that, Squiggle would also do the runtime check to make sure that `7` is in `[5, 10]` range.

The currently supported `f(x: [5, 10]) = ...` syntax for parameters annotation is a syntax sugar on top of domains: `[5, 10]` annotation used after `:` gets upgraded to the domain. This would work in the same way for variable declarations, `x: [5, 10] = 7`.

We will support other annotation shortcuts in the future, because full domain constructors are too verbose. Also, common types will also become domain constructors, e.g. `x: Number = 7`. `Number` is a function, but the "annotation -> domain" syntax sugar layer will upgrade it to the domain, so it's the same as, let's say, `Number.anyValueDomain()`.

This is the very preliminary plan for the future type system in Squiggle. It's going to be somewhat complicated to implement, because it's heavy on generics, but it seems doable. I'm not sure when we'll get around to implementing all this, tbh.

michaeldickens commented:
Come to think of it, I don't think runtime type checking works for type inference. For example, consider

    dist = 5
    otherDist = 3
    velocity :: m/s = dist / otherDist
    x :: m = dist + otherDist

Line 3 does not strictly enforce types on `dist` and `otherDist` so it doesn't complain. Line 4 enforces that `dist` and `otherDist` both have type `m`, which contradicts line 3, but line 3 already ran so the unit type checker doesn't complain.

michaeldickens commented:
For blocks, should it be allowed to re-define a variable inside a block with a new unit type? For example,

    x :: kg = 1
    {
      x :: lbs = 2
      x
    }

This seems more logical given how block scoping rules work, but it could be confusing in some situations, for example:

    x :: kg = 2.2
    y :: lbs = 1
    {
      x = y
    }

You might expect this to raise a type error but it doesn't, because the `x` inside the block is considered distinct from the outer `x`.

---

I already implemented backward unit type inference so it looks like it works (but there could be edge cases I haven't thought of). The way it works is by making a list of constraints, for example

    x :: kg = 1
    y :: lbs = x

produces the list

    x^1 = kg^1
    y^1 = lbs^1
    y^1 = x^1

If you take the logarithm of each of these, you get a system of linear equations, which you can solve using Gaussian elimination (which runs in approximately O(n^3)). In this case, the solution will look something like

    x = kg
    y = lbs
    0 = lbs / kg

The third line is nonsensical which means there's a type error (and you can tell where the type error is by looking at which step in the Gaussian elimination algorithm produced the invalid line).

michaeldickens commented on packages/squiggle-lang/src/ast/peggyParser.peggy:
```diff
@@ -58,13 +58,38 @@ decorator
+
+implicitType
+  = ""
+    { return h.nodeImplicitType(location()); }
```
Making `typeSignature` optional is non-trivial because an empty type has actual semantics and it's useful to have an `ImplicitType` node to handle this. You could also handle it by checking `NodeLetStatement.typeSignature === undefined` but (1) this would require a special case and (2) my preference is to be explicit about when types are implicit.

berekuk replied:
> You could also handle it by checking `NodeLetStatement.typeSignature === undefined` but (1) this would require a special case and (2) my preference is to be explicit about when types are implicit.

I don't understand why undefined check is a problem (also you can use `!NodeLetStatement.typeSignature`, since `typeSignature` is an object and so it's truthy); you either have to do that or check `typeSignature.isImplicit` everywhere, and the latter is just more verbose and less memory-efficient.

michaeldickens replied:
I will change this to work like you suggested, but since it's just a cosmetic change, I'll do it after I get all the functionality working.

My latest commit introduced a similar pattern [here](https://github.com/quantified-uncertainty/squiggle/pull/3323/files#diff-c551d40e63d35b3be13c121195361b9e75f4b55190ffc1db82700fb5f8320f6eR61-R65) where I have an explicitly-marked object that could just be `undefined`, I can change that too if you want.

berekuk commented:
> For blocks, should it be allowed to re-define a variable inside a block with a new unit type?

Yes, I think both examples you give should be allowed.

Even this should be allowed, because Squiggle doesn't have mutable variables, any new declaration is shadowing:

```
x :: kg = 1
x :: lbs = 2
```

(btw something like this is also [allowed in Rust](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=d408f52d843902c53a47479a61cf94c4), so it's not that unusual)

---

> If you take the logarithm of each of these, you get a system of linear equations, which you can solve using Gaussian elimination (which runs in approximately O(n^3)).

Huh.
- it's impressive that you did this
- but I'm still skeptical that it's practical

The examples you give in your last comment are for the simpler case where the entire expression consists of divisions and multiplications.

But in practice you'll also have addition and subtraction (and also other operations and builtin functions); and those seems to map to something SAT-shaped, in general case?

Also, this particular puzzle with unit types is a very special case of a more generic longer-term goal of having type inference in general, and it feels very weird to have it so overpowered for units and underpowered for everything else; it's also hard to predict right now how unit types with backward inference will interact with other type inference steps, the approaches will get in conflict at some point, and so the cost of maintenance of complex solutions seems too high to me.

michaeldickens commented:
> Even this should be allowed, because Squiggle doesn't have mutable variables, any new declaration is shadowing:

Ok, I didn't realize this, re-declarations don't work with my current implementation. But I was going to have to change it to make blocks work anyway so I'll change this too.

> But in practice you'll also have addition and subtraction (and also other operations and builtin functions); and those seems to map to something SAT-shaped, in general case?

`+`,` `-`, `==`, etc. require that the two arguments have the same unit type. This is a pretty easy constraint to check. I've already implemented backward type inference for all (valid[1]) binary operations and it works. From looking thru the list of builtins, I *think* it should work, but I haven't looked at them in detail.

[1] Some binary operations, like `&&`, don't produce unit types. `^` only produces a unit type if the exponent is an integer literal so I just let the result of `^` be any type (I could change this later).

> Also, this particular puzzle with unit types is a very special case of a more generic longer-term goal of having type inference in general, and it feels very weird to have it so overpowered for units and underpowered for everything else; it's also hard to predict right now how unit types with backward inference will interact with other type inference steps, the approaches will get in conflict at some point, and so the cost of maintenance of complex solutions seems too high to me.

Wouldn't unit types and normal types be totally separate systems? Although I can see that it's a weird UX if inference works differently for types vs. unit types.

If it's not too much trouble, **could you write a few test cases that you expect would fail for backward type inference?** I am writing my own test cases but since you're more knowledgeable, you can probably come up with more edge cases than me. If backward type inference turns out not to work, or if we decide forward inference is better UX, then it shouldn't be hard for me to rewrite it to do forward inference only.

michaeldickens commented:
Type checking and backward inference now work (according to my test suite) for blocks. I'm not sure I understand Squiggle's scoping rules but I *believe* the type checker behaves correctly according to the scoping rules.

michaeldickens commented:
I have an idea for how to do type inference on generic functions, I haven't implemented it yet but I believe it will work. Should there also be syntax for explicit generics?

It's not obvious how to do syntax for generic functions because a unit type can be any identifier. A few ideas:

1. generic types are numbers: `f(x :: 1, y :: meters) :: 1 = { ... }`
2. generic types are capitalized words: `f(x :: Any, y :: meters) :: Any = { ... }`
3. generic types are indicated by some kind of syntactical marker, for example square brackets: `f(x :: [any], y :: meters) :: [any] = { ... }`

michaeldickens commented:
BTW my plan is to fully implement backward inference to check that it works, but if you prefer forward-only inference (because it's easier to maintain, or better UX, or whatever), it shouldn't be too hard for me to rewrite to do forward-only inference.

michaeldickens commented:
Having investigated more, it looks like it's difficult to do compile-time type checking for functions whose values get determined at runtime, such as higher-order functions:

```
apply(f, x) = f(x)
double(x) = 2*x
x :: meters = 2
y = apply(double, x)
```

It's theoretically doable but it would take quite a bit of work.

I can see a few ways of implementing unit type checking that avoid this complexity:

1. Do function type checking on a "best-effort" basis: the type checker checks functions when it can, but for functions whose values aren't known at compile time, it just assumes the types are correct.
2. Type check at compile time when possible, and type check everything else at runtime.
3. Type check everything at runtime.

I think number 1 is reasonable because type checking on non-static functions is pretty niche so I doubt many people will miss it. But number 3 is perhaps clearer.

berekuk commented:
Hi Michael! Sorry for the slow response, I was occupied with finishing another big PR.

I'll address the meta first; I realize that communication on this is complicated because:
- I don't know much about your background so I don't know if you have experience with advanced strong-typed languages such as OCaml or Haskell; but my impression is that you're mostly trying to figure it out as you proceed, in ad-hoc fashion
- at the same time, I'm very far from being an expert myself! I know _some_ things, but could easily be wrong about some stuff, and the possible skill ceiling on type system knowledge seems extremely high
- I think some amount of ad-hoc approach in this is fine and unavoidable, but it's also easy to paint yourself into the corner because some early feature (like backward inference) will block future features due to implementation complexity or computational complexity, so it's good to have some foresight into what's easy here and what's hard

---

On backward inference, I've realized there's a trivial example that shows that it's not doable in general:

```
x = 1
y = 2
z :: kg * m = x * y
```

So, `x :: kg; y :: m` or `x :: m; y :: kg`? No way to know.

(I believe this is a _very simple_ example of a much more complicated problem; when you have backward inference and also other features such as functions and generics

---

berekuk commented:
Hi Michael! Sorry for the slow response, I was occupied with finishing another big PR.

I'll address the meta first; I realize that communication on this is complicated because:
- I don't know much about your background so I don't know if you have experience with advanced strong-typed languages such as OCaml or Haskell; but my impression is that you're mostly trying to figure it out as you proceed, in ad-hoc fashion
- at the same time, I'm very far from being an expert myself! I know _some_ things, but could easily be wrong about some stuff, and the possible skill ceiling on type system knowledge seems extremely high
- I think some amount of ad-hoc approach in this is fine and unavoidable, but it's also easy to paint yourself into the corner because some early feature (like backward inference) will block future features due to implementation complexity or computational complexity, so it's good to have some foresight into what's easy here and what's hard

---

On backward inference, I've realized there's a trivial example that shows that it's not doable in general:

```
x = 1
y = 2
z :: kg * m = x * y
```

So, `x :: kg; y :: m` or `x :: m; y :: kg`? No way to know.

(I believe this is a _very simple_ example of a much more complicated problem; when you have backward inference and also other features such as functions and generics

---

berekuk commented:
Hi Michael! Sorry for the slow response, I was busy with finishing another big PR.

I'll address the meta first; I realize that communication on this is complicated because:
- I don't know much about your background so I don't know if you have the experience with advanced strong-typed languages such as OCaml or Haskell; but my impression is that you're mostly trying to figure it out as you proceed, in ad-hoc fashion
- at the same time, I'm very far from being an expert myself! I know _some_ things, but could easily be wrong about some stuff, and the possible skill ceiling on the type system knowledge seems extremely high
- I think some amount of ad-hoc approach in this is fine and unavoidable, but it's also easy to paint yourself into the corner because some early feature (like backward inference) will block future features due to implementation complexity or computational complexity, so it's good to have some foresight into what's easy here and what's hard

---

On backward inference, I've realized there's a trivial example that shows that it's not doable in general:

```
x = 1
y = 2
z :: kg * m = x * y
```

So, `x :: kg; y :: m` or `x :: m; y :: kg`? No way to know.

I haven't figured out your implementation yet, maybe you're doing backward inference only when one of parameters is untyped? Or does this result in error because there's not enough information?

Either way, I believe this is a _very simple_ example of a much more complicated problem; when you have backward inference and also other features such as functions and generics, the complexity will blow up.

---

Another evidence against backward inference is that I can't think of any language that has it.

I think Prolog kinda does, but not on type system level?

And OCaml/Haskell have powerful type inference that can look like backward inference but it's not exactly that (I'll address this in the next section.)

---

Now, on generic functions. There's a long ladder of power that starts with "no types" and ends with "theorem provers", with several steps in between:
1. weakly typed languages — everything is checked at runtime
2. everything up to TypeScript and below — if args are not typed then return values are also not typed; `f(x) = x` always returns `any`
3. OCaml level, i.e. [Hindley-Milner](https://en.wikipedia.org/wiki/Hindley%E2%80%93Milner_type_system) — types are inferred based on operations done in function definition, even if parameters are not typed. Here's [Rescript example](https://rescript-lang.org/try?version=v11.1.2&code=DYUwLgBAZgjBC8EAeCB8yBQHSTHRsAFAKwCUEA9BRAJYB2Y24EYATAtDIQERQD2fbuSoQAzmABO9AOZYc0dohTx0KALToAghIkBDAJ4A6aeACqdUbqghCABlJzmYAMwcorQgG0YAGlY-nAF1hanpGeTAAFjcPT14Bbh8IbgAjXQluYMpqcSk6aSA). This level is very cool, but unfortunately ["Attempts to combine subtyping and type inference have caused quite some frustration."](https://en.wikipedia.org/wiki/Hindley%E2%80%93Milner_type_system#Subtyping), and I'm pretty sure we'll need subtyping in Squiggle.
4. Languages that support dependent types. These can do anything, at the cost of performance and cognitive load for your users. Squiggle will never get to this stage.

Btw, it's not like Hindley-Milner is complicated; the math in wikipedia page or other common explanations may look complicated if you don't know the syntax, but I have [this gist bookmarked](https://gist.github.com/oxyflour/f98432aa400daa225d04), it implements it in ~200 lines of code. The question is more of: should we do it?

RIght now Squiggle is at (1), and we aim to get it to (2) or (3), but my understanding is that subtyping vs type inference tradeoff is key here and it's not trivial. The detailed explanation would take too long (this comment is already too long), so I'll write a dense outline of my thoughts on this and hope that it'll be at least somewhat understandable:
- you can either do subtypes or algebraic data types, but you need one of these
- Squiggle doesn't have custom structs so users pass dicts around and sometimes you'll want to add or remove or transform a dict; similar to TypeScript, a good type engine should infer transformation shapes; this is "subtyping"
- we could add structs, but albegraic data types (GADTs) are more clunky/more formal, e.g. you'll need [Option types](https://en.wikipedia.org/wiki/Option_type) and unwrap things manually, and it's not clear that it's worth it in a language designed to be easy-to-use for quick sketches
- so I guess we'll have to do (2)? but I've been postponing this decision until we introduce custom structs because then it'll be more clear how far we can push this boundary, type inference in polymorphic functions is _really nice_ when you have it, maybe some partial version of it is doable?

---

Anyway, my advice for now (unless you know something that I don't, which is very possible, again, I'm very much not an expert here and can be confused):
- drop backward inference
- do the easiest thing possible with inferring units, I don't care if it's runtime or compile-time
- don't infer return types in functions for now, if you do compile-time; syntax for explicitly marking the return value with units is fine but also optional

michaeldickens commented:
On my background:

- I'm fairly familiar with Haskell, I've written maybe 10,000 lines of code in it. Not familiar with OCaml.
- I've written two different compilers for toy languages, one of which was statically typed (it was object-oriented, not algebraic types).

I'm figuring out unit types as I go because unit types differ from normal types in some important ways:

- you can use a unit type at any time without defining it (you can say `x :: kg = 1` without ever defining what a `kg` is)
- unit types don't have internal structures the way objects or algebraic types do
- unit types don't have typeclasses
- not every value has to have a unit type

It doesn't really make sense to do Hindley-Milner for unit types because there is no such thing as a unit typeclass. (I suppose we could implement unit typeclasses but I don't see much point.) And the complexity of type inference mainly comes from (1) typeclasses and (2) higher-order functions, so if you don't have typeclasses, that makes things simpler. (Higher-order functions are still hard.)

I did look a bit for pre-existing work on how to implement unit types and some work exists but there's no established best practice.

> you can either do subtypes or algebraic data types, but you need one of these

Can you explain your reasoning? I wouldn't think you'd want either of those for unit types.

As to your backward inference example—you're right that there are some other situations where backward inference is hard, but this sort of case isn't a problem:

```
x = 1
y = 2
z :: kg * m = x * y
```

The unit type system imposes a constraint that `x * y` must have type `kg * m`, but it has a degree of freedom, so in the absence of any other constraints, `x` and `y` have unspecified unit types and there is no error.

As of my latest commit:

- Implicitly-typed generic functions work (at least my tests pass, there might be edge cases I haven't found).
- Explicitly-typed functions work.
- Backward inference works.

However:

- There is no syntax for explicitly-typed generic functions.
- Type inference for functions only works when the function's value is known at compile time. For example, the type checker can't infer types on this: `y = (condition ? f : g)(x)`. I have a vague idea of how to fix this but it would require significant rewrites. Right now the type checker just ignores functions whose values aren't known at compile time.

My previous two comments described an implementation that does 'function type checking on a "best-effort" basis', which I have now implemented.

"the easiest thing possible with inferring units" would be backward type inference for functions and variables but excluding functions whose values aren't known at compile time, because I've already implemented that.

If you expect the unit type checker to stumble in certain situations, I think the most helpful thing would be for you to write Squiggle code that you expect to cause issues, and I can turn that code into unit tests. You might also have a look at the tests I've already written, especially the ones starting from [this line](https://github.com/quantified-uncertainty/squiggle/pull/3323/files#diff-76972a70638c99b92a7c9e4b63ebbd4a58dbbdc1a90902850d00c6333c741487R205).

berekuk commented:
> On my background:

Thanks for clarifying!

> It doesn't really make sense to do Hindley-Milner for unit types because there is no such thing as a unit typeclass.
> [...]
> Can you explain your reasoning? I wouldn't think you'd want either of those for unit types.

I missed an important assumption that I'm making: I mostly wasn't talking about unit types, but about type checking for Squiggle in general, because unit types are just a part of that.

Here's the argument:
- unit types are going to be the innermost type parameter of generics, because we'll eventually have arrays of unit-marked numbers, tuples, distributions over units and so on
- with type inference for types in general, types for expressions that are far away from each other can be connected by type inference
- if those types can be generics over units, then by adding features for units, you're also kind of precommitting to support those features for containers-over-units, which is a much more complicated task, which will require you eventually to pierce through all layers of wrapped high-order types to connect the information about units across the entire type graph

I should look at your implementation more carefully, but for now I'm very hesitant because of these future maintenance costs when we eventually try to combine backward inference with generic container types.

---

Another thing that makes me not very excited about backward inference, tbh, is that it just... doesn't seem very useful?

Like, it's clear why forward inference is useful.

And it's clear why ML-style generic functions or typeclasses are useful, e.g. being able to infer `f(x) = x; f(1) /* number */` without manually writing out generics for `f`.

But backward inference for units is neither of that, and in most cases I can think of you will have to name the unit somewhere, and you might as well do it at the _first_ variable. AFAICT you won't save any keystrokes through inference here?

---

One more problem with backward inference is that O(N^3) is kinda slow. Squiggle has to be real-time (by default we recompile and run the code on each keystroke!). I'm not sure if the slowdown will become noticeable around 30 statements or around 100; 100 doesn't seem so bad, but if it's closer to ~30 statements then it's a problem.

michaeldickens commented:
> I mostly wasn't talking about unit types, but about type checking for Squiggle in general, because unit types are just a part of that.

I see. I was thinking of unit types and normal types as separate things. My thinking was that

- unit types are more useful than normal types for Squiggle's applications (quantitative models etc.)
- unit types only make sense for numbers (and number-y types like symbolic dists) so there's no need to make unit types well-defined for all types (strings, etc.)
- unit types don't work the same way as types, for example multiplication preserves type but changes unit type

But if unit types are a subsystem within the general type system, then that makes things more complicated, as you say.

In that case, this PR seems problematic because it encodes a bunch of decisions about the unit type system that the general type system will have to adopt, but I haven't put any thought into whether the design makes sense for a general type system—at a glance, I don't think it does.

At least for how I use Squiggle, unit type checking is much more useful than general type checking so I'd like to see it deployed first, but I wouldn't want to deploy it in a way that could mess things up later down the road. But that makes me think the type system needs to be fully specced out before unit types can be implemented.

If you have a general picture of how types will work, and you can see a way to implement unit types that won't be a problem in the future, I'm happy to implement it. It sounds like you're ok with either (1) compile-time checking with forward inference or (2) run-time checking, but you don't think backward inference will work with general types. I agree, I don't think it will work—it works fine for unit types but then you can't interlock the two type systems.

I'm not sure how to proceed—I want Squiggle to have unit types and I'm happy to implement them, but I don't think I'm qualified to make design decisions about how the general type system should work, since I don't have the broad perspective on Squiggle's design philosophy that you have. I'm sure you're working on lots of other stuff so I don't want to push more work onto you.

If you think my current design will work other than the backward inference, it wouldn't be hard for me to rewrite it to do forward-only inference.

-----

The current design only allows unit-type-checking on statically-defined functions, which could be an issue, and I don't think compile-time checking will work at all for arrays/dicts because right now arrays/dict elements can contain any arbitrary types, so it's impossible to know what they contain at compile time. The type checker would probably need to do something like TypeScript where you can enforce types on particular data structures, but untyped data structures can contain anything.

-----

If unit types are supposed to be a part of the general type system, the syntax I chose might not be ideal, but changing the syntax is pretty easy and doesn't affect the type checking logic at all. I had imagined that the syntax would look like

```
explicitUnitType :: kg = 1
explicitNormalType : float = 2
explicitBothKindsOfTypes : float :: kg = 3
```

But perhaps it would make more sense to do something like

```
explicitUnitType : any[kg] = 1
explicitNormalType : float = 2
explicitBothKindsOfTypes : float[kg] = 3
```

-----

> One more problem with backward inference is that O(N^3) is kinda slow.

I just did a quick performance test. With 100 variables it runs ~instantly (a normal test with a few variables takes 0.5 sec, a test with 100 variables also takes 0.5 sec). With 1000 variables it takes 29 sec. With 200 variables it takes 1 sec (so 0.5 sec after the test setup).

I would also note that the reason it's O(n^3) is because it can infer arbitrarily complicated unit type relationships. For example:

```
a = 0
b = 1
c = 2
x :: m/s = a / b
y :: kg/s = b / c
z :: kg/m = c / a
```

Any two of `(x, y, z)` don't conflict, but all three together create a type error.

There's a simpler version of the type checker that can only find conflicts between pairs of variables (so it would fail to find the type error above), and I think that's doable in O(n^2). Admittedly the above example is a pretty weird situation that will likely hardly ever come up in real code.

berekuk commented:
> In that case, this PR seems problematic because it encodes a bunch of decisions about the unit type system that the general type system will have to adopt, but I haven't put any thought into whether the design makes sense for a general type system—at a glance, I don't think it does.

Yep! Happy to know we're on the same page now.

> At least for how I use Squiggle, unit type checking is much more useful than general type checking so I'd like to see it deployed first

I don't mind doing things in whatever order.

For me, general type checking is mostly useful for editor features — autocompletion and so on. Especially with how `->` operator works, it's nice to hint the allowed methods; also `myDict.foo` autocompletion would be very nice.

Potentially also performance gains, but that will be harder to achieve.

> If you have a general picture of how types will work, and you can see a way to implement unit types that won't be a problem in the future, I'm happy to implement it. It sounds like you're ok with either (1) compile-time checking with forward inference or (2) run-time checking, but you don't think backward inference will work with general types. I agree, I don't think it will work—it works fine for unit types but then you can't interlock the two type systems.

Full type system is definitely out of scope of this PR, and even discussing it would take too long now, so the question is how to finish units without creating future problems.

Yes, I'm fine with either (1) or (2).

> The current design only allows unit-type-checking on statically-defined functions, which could be an issue, and I don't think compile-time checking will work at all for arrays/dicts because right now arrays/dict elements can contain any arbitrary types, so it's impossible to know what they contain at compile time. The type checker would probably need to do something like TypeScript where you can enforce types on particular data structures, but untyped data structures can contain anything.

Yeah, even when Squiggle gets types and generic types, it will still have to be gradual typing and not fully statically inferred. But I think it's fine if we don't worry about that for now.

I also won't mind if this PR ended up not supporting functions at all! Happy to merge even if only basic cases such as `x :: kg = 1; y = x` or `x :: kg = 1; y :: t = 2; z = x / y` are supported.

> I had imagined that the syntax would look like
> 
> ```
> explicitBothKindsOfTypes : float :: kg = 3
> ```
> 
> But perhaps it would make more sense to do something like
> 
> ```
> explicitBothKindsOfTypes : float[kg] = 3
> ```

I hope to start with the former and end up with the latter. Then we deprecate `::` in the future release; this is why I suggested `::`, to reserve `:` for this future gradual upgrade.

michaeldickens commented:
Ok, I think I know what to do with this PR then. I will rewrite it do to forward-only compile-time unit type inference. There's a good amount of outstanding work beyond that, on things like updating the prettier plugin and writing docs, which I will get to next.

My biggest open questions right now are:

- [ ] Right now the type checker throws a compile error when it finds a type error, but it can only throw one error at a time, and the type checker can find all errors simultaneously. Is there some better way of raising errors so the user can see all of them at once?
- [ ] For this PR, do we want Squigglehub to have some UI for displaying unit types and type errors? Or should that be separate? (Might be better to do it separately so this PR doesn't get too big.)
- [ ] Squiggle already uses the word "unit" to refer to multiplicative suffixes, like how "100k" means 100,000. Is it bad if "unit" also refers to unit types? Should they be called something else?

michaeldickens commented:
A couple more notes:

- [ ] Right now, all built-in functions are treated as taking and returning any type. It looked like too much work to add type annotations to each of the hundreds of built-in functions. But that means if we do add type annotations in the future, it won't be backward-compatible.
- [ ] I was thinking about it and I realized that downgrading from backward type inference to forward-only is a non-breaking change, but adding backward type inference is a breaking change. I rewrote the type inferrer to be simpler but it can still (sometimes) do backward type inference. It's easy to change to do forward-only inference—it only requires deleting a single line of code.

michaeldickens commented on packages/prettier-plugin/src/printer.ts:
```diff
@@ -187,6 +188,7 @@ export function createSquigglePrinter(
               softline,
               ")",
             ]),
+            node.value.returnUnitType ? typedPath(node).call(print, "value", "returnUnitType") : "",
```
@berekuk This line of code (and two other similar lines) are giving TypeScript errors:

> Argument of type '(path: AstPath<SquiggleNode>) => Doc' is not assignable to parameter of type 'CallCallback<NodeTypeSignature | undefined, Doc>'.

Apparently I need to coerce the type to non-undefined but this line is doing some weird dynamic function call thing that I don't understand, so I don't know how to fix it. Do you know how to fix this?

michaeldickens replied:
I just put this as `@ts-ignore` for now since it didn't seem that important

michaeldickens commented:
This PR is complete and ready for review, other than the concerns raised in my last two comments. There are still plenty of missing features (as we've discussed already) but I don't think they're necessary for an MVP unit-type system.

I plan to write another PR to implement a display on Squigglehub that shows variables' inferred unit types.

berekuk commented:
As we discussed in Discord, I've merged this without going into too many details in the review, because:
- tests are passing
- AFAICT this feature doesn't break any existing code
- in the next few days I plan to start on the full-blown semantic analysis layer for Squiggle, which implies significant further changes to this code

Thank you for doing this PR, I believe it's the most substantial open-source contribution we've had to Squiggle, by far!
